src/esn.inbox.libs/app/services/mailboxes/mailboxes-service.js

```diff
@@ -253,62 +260,85 @@ angular.module('esn.inbox.libs')
     }
 
+    /**
+     * Only return mailboxes where isShared is true
+     */
     function _getSharedMailboxes(mailboxes) {
       return _.filter(mailboxes, inboxSharedMailboxesService.isShared);
     }
 
-    function _getDifferenceById(toInspect, toExclude) {
-      return _.difference(_.map(toInspect, 'id'), _.map(toExclude, 'id'));
-    }
-
-    function updateSharedMailboxCache() {
-
+    function updateMailboxCache() {
       return withJmapClient(function(jmapClient) {
         return jmapClient.mailbox_changes({
           accountId: null,
           sinceState: inboxMailboxesCache.state
         }).then(function(changes) {
+          if (changes.hasMoreChanges) {
+            return _updateCacheFetchingAllMailboxes(jmapClient);
+          }
           if (changes.newState !== inboxMailboxesCache.state) {
-            return jmapClient.mailbox_get({
-              accountId: null,
-              ids: null
-            })
-              .then(function(mailboxes) {
-                _updateMailBoxCacheState(mailboxes.state);
-
-                return mailboxes.list;
-              })
-              .then(function(mailboxList) {
-                return _addSharedMailboxVisibility(_getSharedMailboxes(mailboxList));
-              })
-              .then(function(sharedMailboxList) {
-                return _updateSharedMailboxList(sharedMailboxList);
-              })
-              .then(function() {
-                $rootScope.$broadcast(INBOX_EVENTS.SHARED_FOLDERS_UPDATED);
-              });
+            return _updateCacheFetchingMailboxChanges(jmapClient, changes);
           }
-
-        }).then(function() {
-          return _getSharedMailboxes(inboxMailboxesCache.list);
         });
       });
     }
 
-    function _updateSharedMailboxList(sharedMailboxList) {
-      const sharedMailboxCache = _getSharedMailboxes(inboxMailboxesCache.list);
-      const removedSharedFoldersIds = _getDifferenceById(sharedMailboxCache, sharedMailboxList);
+    function _updateCacheFetchingMailboxChanges(jmapClient, changes) {
+      _updateMailBoxCacheState(changes.newState);
 
-      if (!_.isEmpty(removedSharedFoldersIds)) {
+      const mailboxesToFetch = _.union(changes.created, changes.updated);
 
-        _removeMailboxesFromCache(removedSharedFoldersIds);
+      if (!_.isEmpty(changes.destroyed)) {
+        _removeMailboxesFromCache(changes.destroyed);
 
-        if (removedSharedFoldersIds.includes($state.params.context) === true) {
+        if (changes.destroyed.includes($state.params.context) === true) {
           $state.go('unifiedinbox.inbox', { type: '', account: '', context: '' }, { location: 'replace' });
         }
+
+        if (_.isEmpty(mailboxesToFetch)) {
+          // in this case the event is not broadcasted by _updateMailboxCache but needed
+          $rootScope.$broadcast(INBOX_EVENTS.FOLDERS_UPDATED);
+        }
+      }
+
+      if (!_.isEmpty(mailboxesToFetch)) {
+        return jmapClient
+          .mailbox_get({
```

Comment:
you don't need to go to a new line here

Reply:
Ok, I don't know about this one, but I fixed it.